protocol_decoder/src/decoding.rs

```diff
@@ -281,8 +298,74 @@ impl ProcessedBlockTrace {
                 let dummy_txn = create_dummy_gen_input(other_data, initial_trie_state, 0);
                 gen_inputs.insert(0, dummy_txn);
             }
-            _ => (),
+            _ => dummies_added = false,
         }
+
+        dummies_added
+    }
+
+    /// The withdrawals are always in the final ir payload. How they are placed
+    /// differs based on whether or not there are already dummy proofs present
+    /// in the IR. The rules for adding withdrawals to the IR list are:
+    /// - If dummy proofs are already present, then the withdrawals are added to
+    ///   the last dummy proof (always index `1`).
+    /// - If no dummy proofs are already present, then a dummy proof that just
+    ///   contains the withdrawals is appended to the end of the IR vec.
+    fn add_withdrawals_to_txns(
+        txn_ir: &mut Vec<TxnProofGenIR>,
+        other_data: &OtherBlockData,
+        final_trie_state: &mut PartialTrieState,
+        withdrawals: Vec<(Address, U256)>,
+        dummies_already_added: bool,
+    ) -> TraceParsingResult<()> {
+        // Withdrawals update balances in the account trie, so we need to do that here.
+        Self::update_trie_state_from_withdrawals(withdrawals.iter(), &mut final_trie_state.state)?;
+
+        match dummies_already_added {
+            false => {
+                // Guaranteed to have a real txn.
+                let final_ir = txn_ir.last().unwrap();
+
+                // Dummy state will be the state after the final txn.
+                let withdrawal_dummy =
+                    create_dummy_gen_input(other_data, final_trie_state, final_ir.txn_idx + 1);
+
+                // If we have no actual dummy proofs, then we create one and append it to the
+                // end of the block.
+                txn_ir.push(withdrawal_dummy);
+            }
+            true => {
+                // If we have dummy proofs (note: `txn_ir[1]` is always a dummy txn in this
+                // case), then this dummy will get the withdrawals.
+                txn_ir[1].gen_inputs.withdrawals = withdrawals;
```

Comment:
I don't think this dummy txn has its `post-state` updated to account for the withdrawals, does it?